### ISSUE TITLE
stream.end() does not return connection to pool - race condition

### ISSUE THREAD

**Issue by rkaw92** [tgriesser/knex#2324]
I'm currently using the stream API on Node v8.1.4 with knex 0.14.0.

It seems that the connection is correctly returned to the pool when I read the entirety of the stream's contents. However, if I don't reach the end of the stream, but instead use `stream.end()` immediately (or do nothing, it really doesn't seem to matter), the connection is never returned to the pool.

This is contrary to [the wiki](https://github.com/tgriesser/knex/wiki/Manually-Closing-Streams), which states that one should call `.end()`  to return the connection to the pool, which it does not seem to be doing in my case.

Consider this trivial example (assuming `CREATE TABLE items (id SERIAL);`, and it **needs to be empty**, or otherwise see *Note 1*):
```js
const db = require('knex')({
	client: 'pg',
	connection: 'postgres://postgres:postgres@127.0.0.1/postgres'
});

const queryStream1 = db('items').select('*').stream();
queryStream1.on('data', function() {});
//.end() commented out for now - everything works as expected
//queryStream1.end();
```

Now, uncomment the last line (`.end()`) - this does not return the connection to the pool. Moreover, it actually *prevents the connection from being returned*!

Upon inspecting `Client_PG#_stream`, it seems that this is a race condition. Here is what I think is happening:
* User calls `.stream()` - this is implemented as `Runner#stream`
  * A stream object is created using `new PassThrough({objectMode: true});`
  * This line runs: `const promise = Promise.using(this.ensureConnection(), function(connection) {`
    * When `ensureConnection()` fulfills, the function will be entered entered and it will call `runner.client.stream(runner.connection, sql, stream, options);` - however, this **does not happen just yet**, as we need to wait for the ensureConnection promise
  * The stream is returned
* The user calls `.end()` on the stream, which causes it to emit event **end** immediately, so it will not be emitted again in the future
* Now, the `runner.client.stream(...)` invocation is actually made, and the underlying dialect's client registers an `.on('end')` handler on the stream (in case of PostgreSQL, it's done [here](https://github.com/tgriesser/knex/blob/0.14.0/src/dialects/postgres/index.js#L187)). However, **it is too late**, as the **end** event has already been emitted and we can no longer detect the end of the stream here.
* Thus, the Promise returned from underlying driver's `_stream` invocation is never resolved nor rejected, and the connection never returns to the pool.

*Note 1*: One odd behavior that can throw off tests is that if you have at least 1 record in the table, the connection is returned to the pool. However, this is caused by a promise **rejection**, not fulfilment - since the stream is already closed and we're trying to pipe to it, a stream error is generated and this in turn runs the [rejecter](https://github.com/tgriesser/knex/blob/0.14.0/src/dialects/postgres/index.js#L187) .

At the moment, I'm not sure what the proper fix would be - one option is to listen to "end" events on the source stream rather than the pipe target (meaning listening on `queryStream` instead of `stream` for pg) in the dialect-specific driver code - this ensures that the user can't mess with the stream before we register our handlers on it.

**Comment by rkaw92** [comment]
P.S. To see pool operations, call the example script with `DEBUG=knex:*`

**Comment by dropdevcoding** [comment]
I also tried it by piping into a http response. Even when doing

**Comment by TonyApuzzo** [comment]
This bug seems to be kinda fixed as of knex 0.20.10.  However, the functionality is still a bit odd. According to the nodejs documentation, .end() is not part of the [readable stream](https://nodejs.org/api/stream.html#stream_class_stream_readable) interface.

The current behavior is:
* Calling .end() on the stream returned by knex does close the stream
* Calling .destroy(err) on the stream causes the stream to be closed and if there is an active promise, it is rejected
* Calling .destroy() with no error does not close the stream and does not resolve the promise

The 3rd situation is probably a new issue that I will create after waiting a bit for someone to chime in on this comment.

**Comment by petslane** [comment]
Does not work. Using knex@0.20.13

My basic code is:
```javascript
const stream = query.stream();
stream.pipe(transformer).pipe(res);
req.on('close', () => {
  // stream.unpipe(transformer);
  // stream.end();
  // stream.destroy();
  // stream.destroy(new Error('bla-bla'));
});
```
knex pool is configured to have max 2 connections. Using MySQL.

Tested by making 3 requests and aborting them in sequence (run,cancel,run,cancel,run,cancel).

| Actions in res.on('close') | Result |
| --- | --- |
| `end()` | MySql Processlist shows first query only. `Rows_sent` will be stuck somewhere in 20000 and about a minute, this query disappears from processlist. Log shows **3** connections acquired and **3** queries made. On request cancel it shows **3** connections released. |
| `destroy()` | MySql Processlist shows first **2** queries only. `Rows_sent` will be stuck somewhere in 20000 and about a minute, these queries disappear from processlist. Log shows **2** connections acquired and **2** queries made. On request cancel it shows **0** connections released. After about 1 minutes `KnexTimeoutError` is thrown once (for 3rd request) |
| `destroy(new Error('bla'))` | MySql Processlist shows first query only. `Rows_sent` will be stuck somewhere in 20000 and about a minute, these queries disappear from processlist. Log shows **3** connections acquired and **3** queries made. On request cancel it shows **3** connections released. |
| | |
| `unpipe(transformer)`+`end()` | Same result as `destroy()` |
| `end()`+`unpipe(transformer)` | Same result as `end()` |
| `unpipe(transformer)`+`destroy()` | Same result as `destroy()` |
| `unpipe(transformer)`+`destroy(new Error('bla'))` | Same result as `destroy(new Error('bla'))` |

with all tests after processlist was clear, 4th request never did start query in MySql.

Only workaround that I found is to not kill stream. So on request close I `.unpipe(transformer)` and `.pipe(noopWriter)` that just accepts stream data, but does nothing with it. So on request close, db query keeps getting data from db, but nothing is done with it, and after it's finished, connection is released to pool and next request can use it. But this is not a acceptable solution as these connections will not be available for incoming requests while running and these queries are made with streams because they are slow and take time and have lot of data.

**Comment by petslane** [comment]
Found solution that works with MySQL.

Instead of using connection pooling provided by `knex`, I used that is provided by `mysql`. Before making query with `knex`, I got connection from pool, passed this connection to quey builder with `.connection()` method. And if I need to stop that query (stream), then I just destroy connection, that will end query stream and `mysql` connection pool will recreate that connection when needed.

Example code:
```typescript
// mysql connection pool
import { createPool, MysqlError, PoolConnection } from 'mysql';

const pool = createPool({
  connectionLimit: DB_POOLING_MAX,
  host: DB_HOST,
  port: DB_PORT,
  user: DB_USER,
  password: DB_PASSWORD,
  database: DB_NAME,
  waitForConnections: false,
});

const getConnection = (): Promise<PoolConnection> =>
  new Promise((resolve, reject) => {
    pool.getConnection((error: MysqlError, connection: PoolConnection) => {
      if (error) {
        return reject(error);
      }
      resolve(connection);
    });
  });

// Somewhere in code
// Making query/stream
this.connection = await getConnection();
return this.query.connection(this.connection).stream();

// To abort that long running query/stream
this.connection.destroy();
```

**Comment by issa-tseng** [comment]
okay. much investigation later, here's what i think is happening here. **i have two possible patches but i need feedback on which to take** because i am not familiar with the practices and patterns in this project. /cc @kibertoad (sorry for the nickflash but i promise i did a lot of digging here.)

first, to refute claims that led me wrong:
1. it has nothing to do with reading or draining the stream. that _happens to work_ because then the normal `end`-event termination process occurs.
2. you definitively want to call `.destroy()` and not `.end()` on the stream (so the wiki page is wrong on this).
3. it's totally fine to pipe straight to an http `Response` stream, and call `.destroy()` on the `Response`. that propagates the stream destruction all the way up to knex internals.
4. you don't need to call `.destroy()` with an error.

the real underlying issue is that the error handling for the stream is insufficient to return the connection to the pool in two different ways.

1. because `.pipe()` is used rather than `pipeline`, and there is a `PassThrough` proxying between the actual query results and the user, it is presently _impossible_ to trip a destroy on the underlying `queryStream`. as documented by nodejs core, when using `.pipe` destroying the PassThrough will _not_ destroy the queryStream that is piping to it.
2. when the queryStream _is_ `destroy`ed (as opposed to ended or errored), there is no handling that will ensure the Promise is either resolved or rejected, which means the `finally` is never tripped, and tarn `release` is never called, and the connection is never returned to the pool, even though the database has happily cancelled the query.

there are two ways i see to patch this, both of which i've tried and seem to work for me.

**the first**: in `lib/runner.js`, modify the `stream(options, handler)` method, replacing the construction of the PassThrough stream with the following:

```
    let queryStream;
    const stream = new PassThrough({
      objectMode: true,
      destroy() {
        if (queryStream) queryStream.destroy();
        queryStream.emit('error', new Error('stream destroyed'));
      }
    });
    stream.on('pipe', (qs) => { queryStream = qs; });
```

this ensures first that the queryStream is destroyed if the PassThrough (which is what's given to the user) is destroyed, and second that when it is destroyed an error is emitted to ensure the rejecter is called.

**the second**: in each client (i tried on `dialects/postgres/index.js` and this worked), use `pipeline()` rather than `.pipe()`.

```
      //queryStream.pipe(stream);
      pipeline(queryStream, stream, rejecter);
```

if the use of pipeline is not desired, then the first solution is desired.

thank you for your attention here!

**Comment by issa-tseng** [comment]
oh. and here is my isolated case i was using to repro:

```
const { PassThrough, pipeline } = require('stream');
const knex = require('knex');

const db = knex({
  client: 'pg',
  connection: { /* etc */ },
  pool: { min: 1, max: 1 }
});

let ctr = 0;
const loop = () => {
  console.log('>> starting req', ++ctr);
  const stream = db.select('*').from('submissions').stream();
  const piped = pipeline(stream, new PassThrough({ objectMode: true }),
    (err) => { console.log('<< pipeline error', err); });

  piped.on('data', (x) => {
    console.log('>> got data. terminating');
    piped.destroy();
    setTimeout(loop, 1000);
  });
};

loop();
```

ideally, this should just keep going. without the patch, it locks up after the first try for me.
knex 0.21.1 (=> tarn 3.0.0)
pg 8.2.1
pg-query-stream 3.1.1

**Comment by lognaturel** [comment]
@rkaw92, @dropdevcoding, @TonyApuzzo, @petslane, @ducaale would any of you be up for taking a look at a proposed fix at #3900 and perhaps share some ideas on how to meaningfully test it? Thanks!

**Comment by kibertoad** [comment]
This should have been addrfessed in 0.21.3. Additional testing and feedback would be most appreciated.

**Comment by issa-tseng** [comment]
the patch as released appears to work against the test case i posted above. that test case is the closest i can i conceive to a test. if i have time i can try to work it into the suite but it'll be a while.

**Comment by Cellule** [comment]
in node internals `_stream_writable.js` it seems to destroy the stream on `finish`
https://github.com/nodejs/node/blob/a4fce32eabf40e6a1a34f384999bf8660b59639a/lib/internal/streams/writable.js#L743-L771

So even in the normal happy path, `stream.on("error", err => {})` is called with `"stream destroyed"`

I haven't managed to setup my environment properly, but it feels like `error` event would be fired in this test case
https://github.com/knex/knex/blob/405b639c0ce0eb1b928b1aef1af9750c8b099fd3/test/integration/query/additional.js#L103-L114

**Comment by Sleepful** [comment]
seems like stream doesn't close its connection ever, and then you get a stream that is already depleted when you try to use it again. Very odd.

**Comment by kobyzskyi** [comment]
This is working in my case:

```js
dbStream.end()
dbStream.destroy()
```

**Comment by nicholaswmin** [comment]
Why are these docs still up: https://github.com/knex/knex/wiki/Manually-Closing-Streams when `stream.end` doesn't actually do this?

If you need a test case or a PR I'll submit one - this is an extremely important point with huge consequences in any serious production installation.

**Comment by nicholaswmin** [comment]
> ```js
> dbStream
> ```

This works

**Comment by mercmobily** [comment]
Hey everyone, especially @nicholaswmin, @petslane, @dropdevcoding, and @issa-tseng who've invested significant time debugging this issue. You're absolutely right to be frustrated - this has been a long-standing critical issue that affects production systems, and our response as maintainers has been inadequate.

After deep analysis of the issue, the PR #3900 fix, and the current codebase, here's where we stand:

1. **The 2020 "fix" was incomplete** - PR #3900 addressed one specific scenario but left multiple edge cases unhandled
2. **The documentation is wrong** - The wiki still incorrectly suggests using `.end()` 
3. **The implementation is fundamentally flawed** - We're using outdated patterns that don't align with modern Node.js stream best practices
4. **Testing is insufficient** - Stream error cases are barely covered in our test suite

**Why This Keeps Happening**

The root issue is architectural:

1. **PassThrough proxy problem**: We use a PassThrough stream as a proxy, but destroying it doesn't properly propagate to the underlying query stream
2. **Connection lifecycle mismanagement**: The connection release logic relies on events that may never fire in error cases
3. **Race conditions**: Multiple timing issues between stream initialization, error handling, and cleanup
4. **Database-specific quirks**: Each database driver (pg, mysql, mssql) handles streaming differently, but we treat them uniformly

**What We Need to Do - Action Plan**

1. **Fix the destroy propagation properly**
```javascript
// In runner.js stream() method
const stream = new Transform({
  objectMode: true,
  transform: (chunk, _, callback) => {
    callback(null, this.client.postProcessResponse(chunk, queryContext));
  },
  destroy(err, callback) {
    // Ensure connection is ALWAYS released
    if (this.connection) {
      this.client.releaseConnection(this.connection);
      this.connection = null;
    }
    // Destroy the source stream if it exists
    if (this._sourceStream && typeof this._sourceStream.destroy === 'function') {
      this._sourceStream.destroy(err);
    }
    callback(err);
  }
});
```

2. **Add proper cleanup on all termination paths**
   - Handle 'close', 'end', 'error', AND 'finish' events
   - Use a single cleanup function with once-only execution guarantee
   - Add timeout-based cleanup as a failsafe

3. **Fix the connection reference issue**
   - Store connection reference on the stream object itself
   - Ensure it's accessible in all cleanup paths

**Phase 2: Modern Stream Implementation (v4.0.0)**

For this one, we will make a separate issue once Phase 1 is done. The steps for Phase 2:

1. **Switch to pipeline() internally**
```javascript
const { pipeline } = require('stream');

// Replace current pipe() usage with pipeline()
pipeline(queryStream, transformStream, (err) => {
  // Guaranteed cleanup on ANY termination
  this.client.releaseConnection(connection);
  if (err) rejecter(err);
  else resolver();
});
```

2. **Support AbortController**
```javascript
// Allow users to pass AbortSignal
stream(options) {
  const { signal, ...streamOptions } = options;
  // ... create stream with signal support
}
```

3. **Implement proper backpressure handling**
   - Respect highWaterMark settings
   - Pause source stream when transform buffer is full

**Phase 3: Comprehensive Testing**

1. **Add stream-specific test suite**
   - Test early termination scenarios
   - Test error propagation
   - Test connection pool exhaustion
   - Test with minimal pool size (1-2 connections)
   - Test HTTP request abortion scenarios

2. **Database-specific tests**
   - PostgreSQL with pg-query-stream
   - MySQL streaming quirks
   - MSSQL streaming behavior

**Phase 4: Documentation Overhaul**

1. **Update the Wiki immediately**
   - Remove incorrect `.end()` advice
   - Add proper `.destroy()` usage
   - Include common pitfall warnings

2. **Add streaming guide to main docs**
   - HTTP response streaming patterns
   - Proper cleanup examples
   - Connection pool implications
   - Error handling best practices

**Immediate Workarounds (For Users Now)**

Until we fix this properly, here's what works:

```javascript
// Option 1: Always destroy with an error
const stream = knex.select('*').from('users').stream();
req.on('close', () => {
  stream.destroy(new Error('Request aborted'));
});

// Option 2: Use petslane's approach - bypass Knex pooling
const connection = await getConnectionFromMySQLPool();
const stream = knex.select('*')
  .from('users')
  .connection(connection)
  .stream();
  
req.on('close', () => {
  connection.destroy(); // MySQL pool will recreate it
});

// Option 3: Ensure stream is fully consumed
const stream = knex.select('*').from('users').stream();
const devNull = new Writable({
  write(chunk, encoding, callback) { callback(); }
});

req.on('close', () => {
  stream.unpipe();
  stream.pipe(devNull); // Consume remaining data
});
```


**How You Can Help**

1. **Test the upcoming fix** - We'll create a beta branch with Phase 1 fixes
2. **Share your production scenarios** - Help us build comprehensive tests
3. **Review the implementation** - Your debugging experience is invaluable
4. **Help with documentation** - Real-world examples from your use cases

This is a top priority bug that affects core functionality. It will be fixed properly this time.

**Related Issues to Address**
- #5654 - Stream hanging with transactions (different but related)
- #6251 - Migration streaming issues
- #5589 - Interrupted streaming query issues
- #5242 - Stream connections not closing on errors

Thank you all for your patience, detailed debugging, and continued engagement.

**Comment by mercmobily** [comment]
(If you guys are still here and willing to help, please let me know)

**Comment by nicholaswmin** [comment]
Hi @mercmobily; great job. 

Look, we got it working at the end and our implementation is in production for about 1.5 years with heavy enough usage that this problem would have been apparent from day 1. (actually 1st minute). 

Would you like me to share the implementation or are you confident with the direction you've set out already? I don't wanna dilute the tracker with our own stuff needlessly, that's why im asking.

**Comment by mercmobily** [comment]
@nicholaswmin No I would ABSOLUTELY LOVE to know what you guys did, with as much detail as you are able to provide. I really do want to switch to pipeline() in the mid term, to be honest, but things must work *now* for most users. I am eager to know if you patched anything, or what configuration you have. I am obviously VERY happy  that you are still around!

Ideally, it would be great if you could also test the pipeline solution. The tricky part is that it might potentially create downtime -- although we can work really hard to prevent it.
